nixnet/database/_ecu.py

```diff
@@ -37,52 +40,113 @@ def __repr__(self):
+        """
         return _props.get_ecu_comment(self._handle)
 
     @comment.setter
     def comment(self, value):
+        # type: (typing.Text) -> None
         _props.set_ecu_comment(self._handle, value)
 
     @property
     def config_status(self):
+        # type: () -> int
+        """int: Returns the ECU object configuration status.
+
+        Configuration Status returns an NI-XNET error code.
+        You can pass the value to the `nxStatusToString` function to
+        convert the value to a text description of the configuration problem.
+
+        By default, incorrectly configured ECUs in the database are not returned from
+        :any:`Cluster.ecus` because they cannot be used in the bus communication.
+        You can change this behavior by setting :any:`Database.show_invalid_from_open` to ``True``.
+        When the configuration status of a ECU becomes invalid after opening the database,
+        the ECU still is returned from :any:`Cluster.ecus`
+        even if :any:`Database.show_invalid_from_open` is ``False``.
+        """
         return _props.get_ecu_config_status(self._handle)
 
     @property
     def dbc_attributes(self):
         # type: () -> _dbc_attributes.DbcAttributeCollection
-        """:any:`nixnet.database._dbc_attributes.DbcAttributeCollection`: Access the ECU's DBC attributes."""
+        """:any:`DbcAttributeCollection`: Access the ECU's DBC attributes."""
         if self._dbc_attributes is None:
             self._dbc_attributes = _dbc_attributes.DbcAttributeCollection(self._handle)
         return self._dbc_attributes
 
     @property
     def name(self):
+        # type: () -> typing.Text
+        """str: Get or set the name of the ECU object.
+
+        Lowercase letters, uppercase letters, numbers,
+        and the underscore (_) are valid characters for the short name.
+        The space ( ), period (.), and other special characters are not supported within the name.
+        The short name must begin with a letter (uppercase or lowercase) or underscore, and not a number.
+        The short name is limited to 128 characters.
+
+        An ECU name must be unique for all ECUs in a cluster.
+
+        This short name does not include qualifiers to ensure that it is unique,
+        such as the database and cluster name.
+        It is for display purposes.
+        """
         return _props.get_ecu_name(self._handle)
 
     @name.setter
     def name(self, value):
+        # type: (typing.Text) -> None
         _props.set_ecu_name(self._handle, value)
 
     @property
-    def rx_frm_refs(self):
-        return _props.get_ecu_rx_frm_refs(self._handle)
-
-    @rx_frm_refs.setter
-    def rx_frm_refs(self, value):
-        _props.set_ecu_rx_frm_refs(self._handle, value)
+    def frames_received(self):
+        # type: () -> typing.Iterable[_frame.Frame]
+        """list of :any:`Frame<_frame.Frame>`: Get or set a list of frames the ECU receives.
```

Comment:
Can you explain to us how ECUs and Frames relate?  Maybe some pseudo code of how this might look?

I think a received example is also sufficient for transmitted

Reply:
At the moment, I don't have any information other than the C API documentation here:

http://zone.ni.com/reference/en-XX/help/372841P-01/TOC40.htm

All I've done here is give the existing properties (added in #52) more user-friendly names (what they are called in the linked help) and have them use Frame objects instead of refs.

I'll ping Jeff L and see if he can give us a little info on how these would be used.

Reply:
@lilesj can you provide any insight as to Ed's question?

Reply:
An ECU in the database represents a specific node on the BUS or in a sense the physical CAN interface. The ECU item logically maps individual frames to specific ECUs, defining which frames are being transmitted and received by the node. In a fully defined network, each frame in the database will be represented in at least two ECUs. One ECU to transmit and one or more to receive the frame.

An important aspect of the ECU is that it allows you to specify node specific properties which are primarily used by LIN but also used in J1939 over CAN. Examples of node specific properties for LIN are "Master?" and "Protocol Version" where J1939 has a "Node Address" and "Node Name".  The driver can then apply these ECU properties to every frame that it transmits to or from the interface.

Reply:
I wrote a little but of code to set ECU frames.

```
with database.Database("NIXNET_example") as db:
    cluster = db.clusters["CAN_Demo_Box_Cluster"]
    ecu = cluster.ecus["ECU_test"]
    frames = list(cluster.frames.values())
    ecu.frames_transmitted = [frames[1], frames[2]]
    ecu.frames_received = [frames[3], frames[4]]
    db.save()
```

That ends up looking like this in the database editor:

![db_ecu_frames](https://user-images.githubusercontent.com/33291082/37797349-8faf13a4-2de7-11e8-8a22-c4a9b7b02a35.PNG)

Reply:
I put together a little code to set ECU frames:

```
with database.Database("NIXNET_example") as db:
    cluster = db.clusters["CAN_Demo_Box_Cluster"]
    ecu = cluster.ecus["ECU_test"]
    frames = cluster.frames
    ecu.frames_transmitted = [frames["SET_BAUD_RATE"], frames["SET_CDB_TRANSMIT"]]
    ecu.frames_received = [frames["SET_FUNC_GEN_FREQ"], frames["SET_FUNC_GEN_OUTPUT"]]
    db.save()
```

That ends up looking like this in the database editor.

![db_ecu_frames](https://user-images.githubusercontent.com/33291082/37797810-ddfde0fc-2de8-11e8-99d5-35c6e06641d8.PNG)